### PR TITLE
`DataSink` additions

### LINKED ISSUE #7779: Add `as_any` and `metrics` to DataSink trait
### Is your feature request related to a problem or challenge?

* `as_any` to allow downcasting to `Any`
*   Add `fn metrics(&self) -> Option<MetricsSet>` to optionally provide metrics (e.g. number of rows inserted, etc.)

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Dandandan**
## Which issue does this PR close?

Closes #7779

### CODE REVIEW

File: datafusion/physical-plan/src/insert.rs [hunk 1]
```diff
@@ -46,6 +47,23 @@ use datafusion_execution::TaskContext;
...
+    ///
+    /// While the values of the metrics in the returned
+    /// [`MetricsSet`] may change as execution progresses, the
+    /// specific metrics will not.
+    ///
+    /// Once `self.execute()` has returned (technically the future is
+    /// resolved) for all available partitions, the set of metrics
+    /// should be complete. If this function is called prior to
+    /// `execute()` new metrics may appear in subsequent calls.
+    fn metrics(&self) -> Option<MetricsSet>;
```
Comment (alamb) [inline datafusion/physical-plan/src/insert.rs:65]: 🤔  I wonder if there is value in making some sort of End to End example for this feature (so we don't accidentally break it in the future0

File: datafusion/physical-plan/src/insert.rs [hunk 2]
```diff
@@ -46,6 +47,23 @@ use datafusion_execution::TaskContext;
 /// output.
 #[async_trait]
 pub trait DataSink: DisplayAs + Debug + Send + Sync {
+    /// Returns the data sink as [`Any`](std::any::Any) so that it can be
+    /// downcast to a specific implementation.
+    fn as_any(&self) -> &dyn Any;
+
+    /// Return a snapshot of the [`MetricsSet`] for this
+    /// [`ExecutionPlan`].
```
Comment (alamb) [inline datafusion/physical-plan/src/insert.rs:55]:
```suggestion
    /// [`DataSink`].
```

File: datafusion/physical-plan/src/insert.rs [hunk 3]
```diff
@@ -46,6 +47,23 @@ use datafusion_execution::TaskContext;
...
+    ///
+    /// While the values of the metrics in the returned
+    /// [`MetricsSet`] may change as execution progresses, the
+    /// specific metrics will not.
+    ///
+    /// Once `self.execute()` has returned (technically the future is
+    /// resolved) for all available partitions, the set of metrics
+    /// should be complete. If this function is called prior to
+    /// `execute()` new metrics may appear in subsequent calls.
```
Comment (alamb) [inline datafusion/physical-plan/src/insert.rs:64]: This appears to be copy/paste from https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#method.metrics

I wonder if we should just leave a doc link (like "Return a snapshot of the [`MetricsSet`] for this [`DataSink`]. See [`ExecutionPlan::metrics()`] for more details

**Review by alamb — APPROVED** [review]

File: datafusion/physical-plan/src/insert.rs [hunk 4]
```diff
@@ -151,6 +162,16 @@ impl FileSinkExec {
...
     }
+
+    /// Returns insert sink
+    pub fn sink(&self) -> &dyn DataSink {
+        self.sink.as_ref()
+    }
+
+    /// Returns the metrics of the underlying [DataSink]
+    pub fn metrics(&self) -> Option<MetricsSet> {
+        self.sink.metrics()
```
Comment (joroKr21) [inline datafusion/physical-plan/src/insert.rs:173]: Ahh this was the wrong place, it should be in `impl ExecutionPlan for FileSinkExec`

**Pull request merged**